## Derivable Judgement 


### 📚 Importing libraries :-

In [60]:
import pandas as pd
import numpy as np
from scipy.stats import chi2_contingency, chi2
from scipy.stats import f
from scipy.stats import t
from scipy.stats import chi2

print("libraries imported succesfully !")

libraries imported succesfully !


### 📂 Load the dataset :- 

In [5]:
df = pd.read_excel("health_records_dataset_500.xlsx")
df.head()

,record_id,age_group,age,weight,gender,region,smoking_status,exercise_frequency,bmi,blood_pressure,diabetes,hypertension,cholesterol_level,glucose_level,visit_date
0,621a53f1-8e7e-4af5-a770-0e0d672c3a6f,46-60,58,59,Other,North,Former Smoker,Never,27.5,118.4,True,False,160.2,180.5,2026-04-09
1,4b9a2c7f-fe05-4f18-ac1e-f4135d676185,18-25,19,56,Male,West,Former Smoker,Rarely,22.7,138.0,False,False,124.8,99.8,2026-05-31
2,885ba761-d1fd-452c-bbc4-cc269d5912f5,60+,66,65,Male,West,Smoker,Rarely,19.3,123.9,True,False,148.0,213.6,2024-07-09
3,7b278557-5e31-4f78-9ba8-345c4888e1b9,46-60,56,78,Female,North,Former Smoker,Rarely,21.7,157.0,True,True,216.5,216.0,2022-12-04
4,f2627f9f-2986-4313-8c75-94c0a5b0e1da,46-60,57,91,Male,South,Smoker,Never,29.1,154.9,False,True,128.2,104.2,2024-11-21


In [6]:
df.tail()

,record_id,age_group,age,weight,gender,region,smoking_status,exercise_frequency,bmi,blood_pressure,diabetes,hypertension,cholesterol_level,glucose_level,visit_date
495,a2d122ba-c9df-4348-bb20-c27e73b5401b,18-25,20,79,Male,East,Smoker,Daily,28.1,137.2,True,False,162.8,137.8,2024-05-08
496,3997a423-8ad6-4029-8eb7-3b56d1214c3b,60+,68,49,Other,South,Smoker,Rarely,13.9,172.6,False,True,199.9,73.7,2023-09-27
497,ea602958-150a-4100-8cd1-074d61e48c8c,60+,64,112,Male,South,Former Smoker,Daily,42.6,122.8,False,False,258.5,84.7,2023-11-16
498,4bf69da2-bd42-452d-b94e-67d68c803e7f,36-45,42,111,Female,West,Smoker,Daily,48.4,121.4,True,False,127.9,197.2,2023-07-13
499,ef05c6d1-0563-4890-b3f7-9b2387589f88,60+,67,84,Female,East,Former Smoker,Rarely,39.3,159.5,False,True,235.0,101.5,2023-12-23


In [7]:
df.describe()

,age,weight,bmi,blood_pressure,cholesterol_level,glucose_level,visit_date
count,500.000000,500.000000,500.000000,500.000000,500.000000,500.000000,500
mean,48.860000,81.396000,27.845400,137.486800,210.067400,147.150400,2024-04-29 11:57:07.200000256
min,18.000000,45.000000,11.600000,95.800000,120.000000,70.000000,2022-05-09 00:00:00
25%,33.000000,61.000000,20.375000,118.175000,167.200000,112.100000,2023-05-27 00:00:00
50%,50.000000,81.000000,26.600000,137.450000,210.500000,148.200000,2024-04-13 00:00:00
75%,65.000000,101.250000,34.000000,156.725000,253.700000,182.200000,2025-04-09 12:00:00
max,80.000000,120.000000,54.300000,179.900000,300.000000,219.900000,2026-06-06 00:00:00
std,18.407228,22.481593,9.395381,23.556279,51.086098,42.759639,NaN


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   record_id           500 non-null    object        
 1   age_group           500 non-null    object        
 2   age                 500 non-null    int64         
 3   weight              500 non-null    int64         
 4   gender              500 non-null    object        
 5   region              500 non-null    object        
 6   smoking_status      500 non-null    object        
 7   exercise_frequency  500 non-null    object        
 8   bmi                 500 non-null    float64       
 9   blood_pressure      500 non-null    float64       
 10  diabetes            500 non-null    bool          
 11  hypertension        500 non-null    bool          
 12  cholesterol_level   500 non-null    float64       
 13  glucose_level       500 non-null    float64       

### 1️⃣ Formulate at least two hypotheses from the dataset. Example:
* H₀: Smoking has no effect on Diabetes prevalence.
* H₁: Smoking affects Diabetes prevalence.


In [16]:
table1 = pd.crosstab(df['smoking_status'], df['diabetes'])
table1

diabetes,False,True
smoking_status,,
Former Smoker,47,115
Non-Smoker,41,114
Smoker,56,127


In [17]:
chi2, p_value, dof, expected = chi2_contingency(table1)
print("Chi-square value:", chi2)
print("P-value:", p_value)

print("---------------------------------------------------------------")

if p_value < 0.05:
    print("Reject H0: Smoking status affects diabetes prevalence.")
else:
    print("Accept H0: Smoking status has no significant effect on diabetes.")

Chi-square value: 0.7099318299211267
P-value: 0.7011973431403489
---------------------------------------------------------------
Accept H0: Smoking status has no significant effect on diabetes.


#### 📝 Interpretation:
The p-value (**0.701**) is much greater than 0.05, and the chi-square value (**0.71**) is far below the critical value (**5.99**). This means there is **no significant relationship** between smoking status and diabetes in this sample.

#### ✅ Conclusion:
We **accept H₀** — smoking status has no significant effect on diabetes prevalence. Diabetes occurs at similar rates regardless of smoking habit.


#### 🧩 Perform ANOVA Test

In [18]:
daily = df[df['exercise_frequency']=="Daily"]['bmi']
weekly = df[df['exercise_frequency']=="Weekly"]['bmi']
rarely = df[df['exercise_frequency']=="Rarely"]['bmi']
never = df[df['exercise_frequency']=="Never"]['bmi']

In [19]:
groups = [daily, weekly, rarely, never]

k = len(groups)        # number of groups

N = sum(len(group) for group in groups)

print("Number of groups:", k)
print("Total observations:", N)

Number of groups: 4
Total observations: 500


In [20]:
all_bmi = pd.concat(groups)

overall_mean = all_bmi.mean()

print("Overall Mean BMI:", overall_mean)

Overall Mean BMI: 27.8454


In [21]:
SSB = 0

for group in groups:
    group_size = len(group)
    group_mean = group.mean()
    
    SSB += group_size * (group_mean - overall_mean)**2

print("Between Group Sum of Squares:", SSB)

Between Group Sum of Squares: 229.8773833204632


In [22]:
SSW = 0

for group in groups:
    group_mean = group.mean()
    
    SSW += sum((group - group_mean)**2)

print("Within Group Sum of Squares:", SSW)

Within Group Sum of Squares: 43818.44203667954


In [23]:
df_between = k - 1

df_within = N - k

print("DF Between:", df_between)
print("DF Within:", df_within)

DF Between: 3
DF Within: 496


In [24]:
MSB = SSB / df_between

MSW = SSW / df_within


print("Mean Square Between:", MSB)

print("Mean Square Within:", MSW)

Mean Square Between: 76.6257944401544
Mean Square Within: 88.34363313846681


In [25]:
F_value = MSB / MSW

print("F Statistic:", F_value)

F Statistic: 0.8673606882349261


In [26]:
p_value = 1 - f.cdf(F_value, df_between, df_within)

print("P-value:", p_value)

alpha = 0.05

print("------------------------------------------------------------")

if p_value < alpha:
    print("Reject H0")
    print("Exercise frequency has a significant effect on BMI.")
    
else:
    print("Accept H0")
    print("Exercise frequency does not have a significant effect on BMI.")

P-value: 0.4578590014927909
------------------------------------------------------------
Accept H0
Exercise frequency does not have a significant effect on BMI.


#### 📝 Interpretation:
The F-value (**0.87**) is below the critical value (**2.62**), and the p-value (**0.458**) is well above 0.05. The differences in mean BMI across exercise groups (Daily, Weekly, Rarely, Never) are **not statistically significant**.

#### ✅ Conclusion:
We **accept H₀** — exercise frequency does not have a significant effect on BMI in this dataset.


### 2️⃣ Calculate Confidence Intervals for key numerical data like age, weight, etc.

In [27]:
columns = [
    'age',
    'weight',
    'blood_pressure',
]

In [28]:
confidence = 0.95

for column in columns:
    
    data = df[column]
    
    # Number of observations
    n = len(data)
    
    # Mean
    mean = data.mean()
    
    # Standard deviation
    std = data.std()
    
    # Standard error
    standard_error = std / np.sqrt(n)
    
    # t critical value
    t_value = t.ppf((1 + confidence) / 2, n-1)
    
    # Margin of error
    margin_error = t_value * standard_error
    
    # Confidence interval
    lower_limit = mean - margin_error
    upper_limit = mean + margin_error
    
    print("\nColumn:", column)
    print("Mean:", round(mean,2))
    print("95% Confidence Interval:",
          round(lower_limit,2),
          "to",
          round(upper_limit,2))


Column: age
Mean: 48.86
95% Confidence Interval: 47.24 to 50.48

Column: weight
Mean: 81.4
95% Confidence Interval: 79.42 to 83.37

Column: blood_pressure
Mean: 137.49
95% Confidence Interval: 135.42 to 139.56


#### 📝 Interpretation:
We are **95% confident** the true population mean lies within these ranges: Age between **47.24 and 50.48**, Weight between **79.42 and 83.37**, Blood Pressure between **135.42 and 139.56**.

#### ✅ Conclusion:
Since all intervals are fairly narrow, the estimates are **precise and reliable**. The average age is approximately **49 years**, weight approximately **81 kg**, and blood pressure approximately **137**, with only minor variation.


###  3️⃣ Find the Critical Value and p-value to interpret the test results.


In [29]:
from scipy.stats import f

alpha = 0.05

df_between = 3
df_within = len(df)-4

critical_value = f.ppf(
    1-alpha,
    df_between,
    df_within
)

p_value = 1 - f.cdf(
    F_value,
    df_between,
    df_within
)

print("Critical Value:", critical_value)
print("P-value:", p_value)

print("-------------")

if p_value < alpha:
    print("Reject H0")
else:
    print("Accept H0")

Critical Value: 2.6228791395072832
P-value: 0.4578590014927909
-------------
Accept H0


#### 📝 Interpretation:
Using the critical value approach confirms the same result: F (**0.87**) < critical value (**2.62**), and p (**0.458**) > 0.05.

#### ✅ Conclusion:
We **accept H₀** — this re-confirms exercise frequency has no significant effect on BMI.


### 4️⃣ Perform z-test or t-test based on sample size (mean comparison across groups).


In [30]:
male_bmi = df[df['gender']=="Male"]['bmi']

female_bmi = df[df['gender']=="Female"]['bmi']

In [31]:
n1 = len(male_bmi)
n2 = len(female_bmi)

print("Male sample size:", n1)
print("Female sample size:", n2)

Male sample size: 166
Female sample size: 164


In [32]:
mean1 = male_bmi.mean()

mean2 = female_bmi.mean()

print("Male BMI Mean:", mean1)

print("Female BMI Mean:", mean2)

Male BMI Mean: 29.36566265060241
Female BMI Mean: 27.61890243902439


In [33]:
var1 = male_bmi.var()

var2 = female_bmi.var()

In [34]:
standard_error = np.sqrt(
    (var1/n1) + (var2/n2)
)

In [35]:
t_value = (mean1 - mean2) / standard_error

print("T Statistic:", t_value)

T Statistic: 1.6720063978528295


In [36]:
df_value = n1 + n2 - 2

print("Degrees of Freedom:", df_value)

Degrees of Freedom: 328


In [37]:
p_value = 2 * (1 - t.cdf(abs(t_value), df_value))

print("P-value:", p_value)

P-value: 0.09547668207877202


In [39]:
alpha = 0.05

if p_value < alpha:
    print("Reject H0")
    print("There is a significant difference between male and female BMI.")
    
else:
    print("Accept H0")
    print("There is no significant difference between male and female BMI.")

Accept H0
There is no significant difference between male and female BMI.


#### 📝 Interpretation:
The p-value (**0.095**) is greater than 0.05, and the t-value (**1.67**) is below the critical value (**1.97**). The slightly higher male BMI (**29.37**) vs female BMI (**27.62**) is **not statistically significant**.

#### ✅ Conclusion:
We **accept H₀** — there is no significant difference in BMI between males and females.


### 5️⃣ Conduct a chi-square test on categorical data (e.g., Smoking habit vs Disease).

In [61]:
table = pd.crosstab(
    df['smoking_status'],
    df['diabetes']
)

table

diabetes,False,True
smoking_status,,
Former Smoker,47,115
Non-Smoker,41,114
Smoker,56,127


In [62]:
row_total = table.sum(axis=1)

column_total = table.sum(axis=0)

total = table.sum().sum()


expected = np.outer(row_total, column_total) / total


expected = pd.DataFrame(
    expected,
    index=table.index,
    columns=table.columns
)


expected

diabetes,False,True
smoking_status,,
Former Smoker,46.656,115.344
Non-Smoker,44.640,110.360
Smoker,52.704,130.296


In [63]:
chi_square_value = ((table - expected)**2 / expected).sum().sum()

print("Chi-square value:", chi_square_value)

Chi-square value: 0.7099318299211266


In [43]:
rows = table.shape[0]

columns = table.shape[1]


df_value = (rows-1)*(columns-1)


print("Degrees of Freedom:", df_value)

Degrees of Freedom: 2


In [67]:
p_value = chi2.sf(
    chi_square_value,
    df_value
)

In [65]:
alpha = 0.05
critical_value = chi2.ppf(
    1-alpha,
    df_value
)


print("Critical Value:", critical_value)

Critical Value: 5.991464547107979


In [66]:
if p_value < alpha:
    
    print("Reject H0")
    print("Smoking status and diabetes are significantly related.")

else:
    
    print("Accept H0")
    print("Smoking status and diabetes are independent.")

Accept H0
Smoking status and diabetes are independent.


#### 📝 Interpretation:
Same as the earlier chi-square result: chi-square (**0.71**) is far below the critical value (**5.99**), p-value (**0.701**) is well above 0.05.

#### ✅ Conclusion:
We **accept H₀** — smoking status and diabetes are independent of each other.


###  6️⃣ Perform an ANOVA test to check if age groups significantly differ in disease rate.

In [48]:
df['disease_rate'] = df['diabetes'].astype(int)

In [49]:
age_groups = [
    df[df['age_group']=="18-25"]['disease_rate'],
    df[df['age_group']=="26-35"]['disease_rate'],
    df[df['age_group']=="36-45"]['disease_rate'],
    df[df['age_group']=="46-60"]['disease_rate'],
    df[df['age_group']=="60+"]['disease_rate']
]

In [50]:
k = len(age_groups)

N = sum(len(group) for group in age_groups)


print("Number of groups:", k)

print("Total observations:", N)

Number of groups: 5
Total observations: 500


In [51]:
all_data = pd.concat(age_groups)

overall_mean = all_data.mean()


print("Overall Disease Rate:", overall_mean)

Overall Disease Rate: 0.712


In [52]:
SSB = 0

for group in age_groups:
    
    n = len(group)
    
    group_mean = group.mean()
    
    SSB += n * (group_mean - overall_mean)**2


print("SSB:", SSB)

SSB: 1.228554700554701


In [53]:
SSW = 0


for group in age_groups:
    
    group_mean = group.mean()
    
    SSW += sum((group - group_mean)**2)


print("SSW:", SSW)

SSW: 101.2994452994453


In [54]:
df_between = k-1

df_within = N-k


print(df_between)

print(df_within)

4
495


In [55]:
MSB = SSB / df_between

MSW = SSW / df_within


print("MSB:", MSB)

print("MSW:", MSW)

MSB: 0.30713867513867527
MSW: 0.20464534403928342


In [56]:
F_value = MSB / MSW


print("F-value:", F_value)

F-value: 1.5008339260321377


In [57]:
alpha = 0.05


critical_value = f.ppf(
    1-alpha,
    df_between,
    df_within
)


print("Critical Value:", critical_value)

Critical Value: 2.389947844458197


In [58]:
p_value = 1 - f.cdf(
    F_value,
    df_between,
    df_within
)


print("P-value:", p_value)

P-value: 0.20070296022905243


In [59]:
if p_value < alpha:
    
    print("Reject H0")
    print("Disease rate differs significantly among age groups.")

else:
    
    print("Accept H0")
    print("Disease rate does not differ significantly among age groups.")

Accept H0
Disease rate does not differ significantly among age groups.


#### 📝 Interpretation:
The F-value (**1.50**) is below the critical value (**2.39**), and the p-value (**0.201**) is greater than 0.05. Although diabetes rate ranges from **66.7% to 80.8%** across age groups, this spread is **not statistically significant**.

#### ✅ Conclusion:
We **accept H₀** — diabetes rate does not differ significantly across age groups.


### 7️⃣ Calculate Covariance and Correlation between continuous variables (e.g., Age vs. BMI)

In [87]:
age = df['age']
bmi = df['bmi']

In [88]:
age_mean = age.mean()
bmi_mean = bmi.mean()

print("Mean Age:", age_mean)
print("Mean BMI:", bmi_mean)

Mean Age: 48.86
Mean BMI: 27.8454


In [89]:
n = len(df)


covariance = sum(
    (age - age_mean) *
    (bmi - bmi_mean)) / (n-1)


print("Covariance:", covariance)

Covariance: 18.87690981963928


In [90]:
age_std = age.std()

bmi_std = bmi.std()

In [91]:
correlation = covariance / (age_std * bmi_std)


print("Correlation:", correlation)

Correlation: 0.1091510968925344


#### 📝 Interpretation:
The correlation coefficient (**r = 0.109**) indicates a **very weak positive relationship** between age and BMI — age explains only about 1% of the variation in BMI.

#### ✅ Conclusion:
Age and BMI are **essentially unrelated** in this dataset; growing older does not meaningfully predict higher BMI here.
